searx/webadapter.py

```diff
@@ -0,0 +1,234 @@
+import typing
+from searx.exceptions import SearxParameterException
+from searx.query import RawTextQuery, VALID_LANGUAGE_CODE
+from searx.engines import categories, engines
+from searx.search import SearchQuery, EngineRef
+from searx.preferences import Preferences
+
+
+# remove duplicate queries.
+# FIXME: does not fix "!music !soundcloud", because the categories are 'none' and 'music'
+def deduplicate_engineref_list(engineref_list: typing.List[EngineRef]) -> typing.List[EngineRef]:
+    engineref_dict = {q.category + '|' + q.name: q for q in engineref_list}
+    return engineref_dict.values()
+
+
+def validate_engineref_list(engineref_list: typing.List[EngineRef], preferences: Preferences):
+    """
+    Validate query_engines according to the preferences
+
+        Returns:
+            list of existing engines with a validated token
+            list of unknown engine
+            list of engine with invalid token according to the preferences
+    """
+    valid = []
+    unknown = []
+    no_token = []
+    for engineref in engineref_list:
+        if engineref.name not in engines:
+            unknown.append(engineref)
+            continue
+
+        engine = engines[engineref.name]
+        if not preferences.validate_token(engine):
+            no_token.append(engineref)
+            continue
+
+        valid.append(engineref)
+    return valid, unknown, no_token
+
+
+def parse_pageno(form: typing.Dict[str, str]) -> int:
```

Comment:
Wouldn't it be better to add the functionality of the below functions to `preferences.py` somehow?

Reply:
TLDR: 
* Yes, it would make sense to avoid parsing the parameters at two different locations in the code.
* currently webadapter.py is more or less cut / paste from search.py (with the exception of get_selected_categories).
* the parsing from webadapter/preferences/webapp is not easy to follow.
* Perhaps it would be easier to do that in another PR ?

----

There are some differences between webadapter.py and preferences.py: 
* The parameters in the preferences.py are stored in the cookies, and at least one parameter doesn't make sense here: ```pageno```.
* These two parameters could be set in the preferences:
  * ```time_range``` with a default of ```None```
  * ```timeout``` from query (see https://github.com/searx/searx/pull/1640 ), with a default of ```None```

webadapter.py or preferences.py could convert the ```categorie_*=on|off``` parameters into ```categories=...,...```.

preferences.py and webadapter.py parses safesearch in the same way.

------

Example of complicated parsing: the language parameter.

* preferences.py checks the language value [using the searx.utils.match_language function](https://github.com/searx/searx/blob/21dbc7e852a5c48097c0067e78f0600972e54823/searx/preferences.py#L164-L168)
* webapp.py checks the language value from the URL using a regex. This is on purpose, see https://github.com/searx/searx/commit/fd65c1292179fb082e965a1ee6a88b9298a54fc1

Knowing that, let's follow how the language value is set.

https://github.com/searx/searx/blob/21dbc7e852a5c48097c0067e78f0600972e54823/searx/preferences.py#L323-L326

The default value is ```settings['search'].get('default_lang', '')```

searx.webapp.pre_request:
1. initialize a Preferences object.
1. then parses [the cookies](https://github.com/searx/searx/blob/21dbc7e852a5c48097c0067e78f0600972e54823/searx/webapp.py#L458)
1. then parses [the ```preferences``` query parameter](https://github.com/searx/searx/blob/21dbc7e852a5c48097c0067e78f0600972e54823/searx/webapp.py#L470)
1. then parses [the query parameters](https://github.com/searx/searx/blob/21dbc7e852a5c48097c0067e78f0600972e54823/searx/webapp.py#L473)
1. then if the ```language``` value is empty or None, the value is set to the result of ```_get_browser_or_settings_language```:
    * the Accept-Language header value
	* or ```settings['search'].get('default_lang', 'en')``` (second time usage of the settings)

searx.webapp.index calls get_search_query_from_webapp:
  1. it gets the value from ```raw_text_query.languages[-1]```
  1. if the value is empty/not set, it uses ```form.get('language')```: actually the value should already be in the ```preferences``` (```searx.webapp.pre_request```), however the checking is less restrictive here (```VALID_LANGUAGE_CODE.match(query_lang)``` instead of ```searx.utils.match_language```).
  1. if not, from the preferences